### PR TITLE
fix: Nullable parameter and return types in functions/actions

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Close: #1439 

We die not consider the `Nullable` property for function and action imports properly.
In the PR I set all parameters and return to `Nullable=false` to minimize the number of changes in the code.
For type E2E testing I created new actions and function in V2 which are used in type tests in the end.
The method `isNullableProperty` is used to set the default value in the methods.

### CODE REVIEW

File: packages/core/test/test-util/test-services/v2/test-service/function-imports.ts [hunk 1]
```diff
@@ -217,7 +217,7 @@ export function testFunctionImportSharedEntityReturnTypeCollection(
 ): Omit<
   FunctionImportRequestBuilderV2<
     TestFunctionImportSharedEntityReturnTypeCollectionParameters,
-    never
+    never[]
```
Comment (marikaner) [inline packages/core/test/test-util/test-services/v2/test-service/function-imports.ts:220]: Is this on purpose?

File: packages/generator/src/action-function-import/action.ts [hunk 2]
```diff
@@ -7,15 +7,29 @@ import { additionalDocForEntityNotDeserializable } from './function';
...
 
+function actionImportReturnType(actionImport: VdmActionImport) {
+  let type = actionImport.returnType.returnType;
+
+  if (actionImport.returnType.isCollection) {
+    type = `${type}[]`;
+  }
+  if (actionImport.returnType.isNullable) {
+    type = `${type}|null`;
+  }
```
Comment (marikaner) [inline packages/generator/src/action-function-import/action.ts:18]: Minor, but for better readability:
```suggestion
  if (actionImport.returnType.isCollection) {
    type = `${type}[]`;
  }
  
  if (actionImport.returnType.isNullable) {
    type = `${type} | null`;
  }
```

File: packages/generator/src/action-function-import/action.ts [hunk 3]
```diff
@@ -7,15 +7,29 @@ import { additionalDocForEntityNotDeserializable } from './function';
 
 const parameterName = 'parameters';
 
+function actionImportReturnType(actionImport: VdmActionImport) {
+  let type = actionImport.returnType.returnType;
+
+  if (actionImport.returnType.isCollection) {
+    type = `${type}[]`;
```
Comment (marikaner) [inline packages/generator/src/action-function-import/action.ts:14]: I guess this is where `never[]` comes from?

File: packages/generator/src/action-function-import/function.ts [hunk 4]
```diff
@@ -6,19 +6,37 @@ import { getRequestBuilderArgumentsBase } from './request-builder-arguments';
...
+  oDataVersion: string
+) {
+  let type = functionImport.returnType.returnType;
+
+  if (functionImport.returnType.isCollection) {
+    type = `${type}[]`;
+  }
+  if (functionImport.returnType.isNullable) {
+    type = `${type}|null`;
+  }
```
Comment (marikaner) [inline packages/generator/src/action-function-import/function.ts:20]: Same:
```suggestion
  if (functionImport.returnType.isCollection) {
    type = `${type}[]`;
  }
  
  if (functionImport.returnType.isNullable) {
    type = `${type} | null`;
  }
```

Comment (marikaner) [reply packages/generator/src/action-function-import/function.ts:20]: Please note that this function is duplicate except for one word. Maybe there is potential to reduce duplication?

File: packages/generator/src/action-function-import/parameters-interface.spec.ts [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
-import { orderBreakfast } from '../../test/test-util/data-model';
+import {
+  orderBreakfast,
+  orderBreakfastNullable
+} from '../../test/test-util/data-model';
 import { parametersInterface } from './parameters-interface';
 
 describe('parameters-interface', () => {
-  it('functionImportParametersInterface', () => {
+  it('functionImportParametersInterface not Nullable', () => {
```
Comment (marikaner) [inline packages/generator/src/action-function-import/parameters-interface.spec.ts:10]:
```suggestion
  it('functionImportParametersInterface not nullable', () => {
```

File: packages/generator/src/action-function-import/parameters-interface.spec.ts [hunk 6]
```diff
@@ -13,6 +16,25 @@ describe('parameters-interface', () => {
...
+          docs: [`${unixEOL}Breakfast includes a honey toast`]
+        }
+      ],
+      docs: [
+        `${unixEOL}Type of the parameters to be passed to [[orderBreakfast]].`
+      ]
+    });
+  });
+
+  it('functionImportParametersInterface Nullable', () => {
```
Comment (marikaner) [inline packages/generator/src/action-function-import/parameters-interface.spec.ts:29]:
```suggestion
  it('functionImportParametersInterface nullable', () => {
```

File: packages/generator/src/action-function-import/parameters-interface.spec.ts [hunk 7]
```diff
@@ -13,6 +16,25 @@ describe('parameters-interface', () => {
...
+    expect(parametersInterface(orderBreakfastNullable)).toEqual({
+      kind: StructureKind.Interface,
+      name: 'Params',
+      isExported: true,
+      properties: [
+        {
+          name: 'withHoneyToast',
+          type: 'boolean|null',
```
Comment (marikaner) [inline packages/generator/src/action-function-import/parameters-interface.spec.ts:37]:
```suggestion
          type: 'boolean | null',
```

File: packages/generator/src/edmx-parser/v2/edmx-parser.ts [hunk 8]
```diff
@@ -35,8 +35,17 @@ export function parseAssociationSets(root: any): EdmxAssociationSet[] {
...
+  const functionImports: EdmxFunctionImport[] = getPropertyFromEntityContainer(
+    root,
+    'FunctionImport'
+  ).map(f => ({
     ...f,
     Parameter: forceArray(f.Parameter)
   }));
+  return functionImports.map(f => {
+    f.Parameter.forEach(p => {
+      p.Nullable = !p.Nullable ? 'true' : p.Nullable;
```
Comment (marikaner) [inline packages/generator/src/edmx-parser/v2/edmx-parser.ts:47]: I think we already have a function for this. Could you please check if this is applicable here?

File: packages/generator/src/edmx-parser/v2/edmx-parser.ts [hunk 9]
```diff
@@ -35,8 +35,17 @@ export function parseAssociationSets(root: any): EdmxAssociationSet[] {
...
+    });
+    return f;
+  });
```
Comment (marikaner) [inline packages/generator/src/edmx-parser/v2/edmx-parser.ts:50]: Consider nesting `map` twice and maybe more speaking names

```suggestion
  return functionImports.map(functionImport => ({
    ...functionImport
    Parameter: functionImport.Parameter.map(parameter => ({
      ...parameter,
      Nullable: !parameter.Nullable ? 'true' : parameter.Nullable
    }))
  }));
```

File: packages/generator/src/edmx-parser/v4/edmx-parser.ts [hunk 10]
```diff
@@ -69,14 +69,29 @@ export function parseActionImport(root: any): EdmxActionImport[] {
...
+  ).map(actionOrFunction => ({
+    ...actionOrFunction,
+    Parameter: forceArray(actionOrFunction.Parameter),
+    IsBound: false
+  }));
+  return actionsOrFuntions.map(f => {
+    if (f.ReturnType) {
+      f.ReturnType.Nullable = !f.ReturnType.Nullable
+        ? 'true'
+        : f.ReturnType.Nullable;
```
Comment (marikaner) [inline packages/generator/src/edmx-parser/v4/edmx-parser.ts:88]: Same here, please see my comments on the v2 edmx-parser.

File: packages/generator/src/edmx-to-vdm/v4/action-import.spec.ts [hunk 11]
```diff
@@ -259,7 +260,7 @@ function createServiceWithActions(): ServiceMetadata {
   const actionWithReturnWithParameter = createAction(
     'ActionWithReturnWithParameter',
     'TestEntityType',
-    [{ Name: 'StringPara', Type: 'Edm.String' }]
+    [{ Name: 'StringPara', Type: 'Edm.String', Nullable: 'false' }]
```
Comment (marikaner) [inline packages/generator/src/edmx-to-vdm/v4/action-import.spec.ts:263]: "Para" is a bit weird even for a test, consider renaming ;)

File: test-packages/type-tests/test/v4/function-import.spec.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  nonNullable: ''
+});
+
+// $ExpectType FunctionImportRequestBuilder<TestFunctionImportNullableTestParameters, string[] | null>
+testFunctionImportNullableTest({
+  nonNullable: 'someValue'
+});
+
+// $ExpectType Promise<string[] | null>
+const result = testFunctionImportNullableTest({
```
Comment (marikaner) [inline test-packages/type-tests/test/v4/function-import.spec.ts:19]:
```suggestion
testFunctionImportNullableTest({
```

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+For oData v2 and v4 the function import parameter may contain the nullable property e.g.
+
+```XML
+<Function Name="TestFunctionImportNullableTest">
+<Parameter Name="NullableExplicit" Type="Edm.String" Nullable="true" />
+</Function>
+```
+
+In addition for oData v4 the return type is a also not just a string on the `<Function>` tag but a proper lines beeing nullable:
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:19]: "beeing" -> "being"
I don't understand this sentence... "proper lines"?

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+Consider `Nullable` property on function import parameters and return types
+
+## Status
+
+accepted
+
+## Background
+
+For oData v2 and v4 the function import parameter may contain the nullable property e.g.
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:11]:
```suggestion
For OData v2 and v4 the function import parameter may contain the nullable property e.g.
```

Please also fix in the remainder of this document.

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+```XML
+<Function Name="TestFunctionImportNullableTest">
+  <ReturnType Type="Collection(Edm.String)" Nullable="false"/>
+</Function>
+```
+
+## Context
+
+The generator should consider this properties and make the types accordingly and make nullable parameters as:
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:29]:
```suggestion
The generator should generate nullable properties for "Nullable" parameters:
```

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+```ts
+{
+nullableParameter?: string | null
+}
+```
+
+For return types in v4 the type should be `T | null` where `T` is return type.
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:37]:
```suggestion
For "Nullable" return types in v4 the type should be `T | null`.
```

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+The generator did not consider the property corretly which was reported in [this issue](https://github.com/SAP/cloud-sdk-js/issues/1439)
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:38]:
```suggestion
```

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+The default value is also set for the return type for oData v4.
+
+## Decision
+
+Accepted, because `string` typed parameter is better than `number` typed, when building filters.
+
+## Consequences
+
+- When there are real use cases about underlying types and values, it becomes easier to implement them as we parse these information to our VDM models.
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:60]: This seems unrelated (stolen from enum?)...

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+nullableParameter?: string | null
+}
+```
+
+For return types in v4 the type should be `T | null` where `T` is return type.
+The generator did not consider the property corretly which was reported in [this issue](https://github.com/SAP/cloud-sdk-js/issues/1439)
+
+## Alternatives
+
+### Option A
```
Comment (marikaner) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:42]: Seems like this rather belongs into the consequence section? If there is no other option why list options?

**Review by marikaner — CHANGES_REQUESTED** [review]
Code-wise looks good to me, I just had minor comments.
I think you forgot to add a changelog, especially because we need some compatibility notes here.

File: packages/generator/src/edmx-parser/common/edmx-types.ts [hunk 20]
```diff
@@ -30,7 +30,7 @@ export interface EdmxDocumented {
 
 export interface EdmxParameter extends EdmxDocumented, EdmxNamed {
   Type: string;
-  Nullable?: string;
+  Nullable: string;
```
Comment (jjtang1985) [inline packages/generator/src/edmx-parser/common/edmx-types.ts:33]: Is this optional?

File: packages/generator/src/action-function-import/action.ts (hunk 3 above)
Comment (jjtang1985) [inline packages/generator/src/action-function-import/action.ts:14]: question: when it's collection, it's implicitly non-nullable?

Comment (jjtang1985) [reply packages/generator/src/action-function-import/action.ts:14]: same question to function as well

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md (hunk 19 above)
Comment (jjtang1985) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:42]: No needs to mark ABCD for only one option?

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 21]
```diff
@@ -0,0 +1,60 @@
+## Title
+
+Consider `Nullable` property on function import parameters and return types
```
Comment (jjtang1985) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:3]: also mention action?

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md (hunk 18 above)
Comment (jjtang1985) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:60]: copy paste left over?

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+We implemented the following for v2 and v4:
+
+- In the edmx-parser.ts the default value of `Nullable="true"` is set for all parameters of function and action imports.
+- This is passed to the `edmx-to-vdm.ts` and cast to boolean.
+
+The default value is also set for the return type for oData v4.
+
+## Decision
+
+Accepted, because `string` typed parameter is better than `number` typed, when building filters.
```
Comment (jjtang1985) [inline knowledge-base/adr/0023-nullable-function-action-import-parameter.md:56]: copy paste left over?

File: packages/core/test/test-util/test-services/v2/test-service/function-imports.ts (hunk 1 above)
Comment (FrankEssenberger) [reply packages/core/test/test-util/test-services/v2/test-service/function-imports.ts:220]: I think we had a bug in the code missing the `[]` since a collection is returen I think this is correct

File: packages/generator/src/action-function-import/action.ts (hunk 3 above)
Comment (FrankEssenberger) [reply packages/generator/src/action-function-import/action.ts:14]: yes

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md (hunk 19 above)
Comment (FrankEssenberger) [reply knowledge-base/adr/0023-nullable-function-action-import-parameter.md:41]: ups, yes copy paste error. Saw the empty template too late.

File: packages/generator/src/edmx-to-vdm/v2/function-import.ts [hunk 23]
```diff
@@ -39,7 +39,9 @@ export function generateFunctionImportsV2(
           ),
           httpMethod,
           returnType: parseFunctionImportReturnTypes(
-            f.ReturnType,
+            f.ReturnType
+              ? { Type: f.ReturnType, Nullable: 'false' }
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply packages/generator/src/edmx-to-vdm/v2/function-import.ts:43]: for oData v2 the return type is non a XML entry like the parameters but it is given as a property on the `<FunctionImport>` I adjusted the ADR to explain that better

File: packages/generator/src/edmx-parser/v4/edm-types.ts [hunk 24]
```diff
@@ -44,11 +44,16 @@ export interface EdmxActionImport extends EdmxNamed, EdmxNamespaced {
...
 export interface EdmxFunction extends EdmxNamed, EdmxNamespaced {
-  ReturnType?: { Type: string };
+  ReturnType?: EdmxReturnType;
   Parameter: EdmxParameter[];
   IsBound: boolean;
 }
 
+export interface EdmxReturnType {
+  Type: string;
+  Nullable: string;
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply packages/generator/src/edmx-parser/v4/edm-types.ts:54]: In the edmx parser I always set the value to the default `true`

File: packages/generator/src/edmx-parser/common/edmx-types.ts (hunk 20 above)
Comment (FrankEssenberger) [reply packages/generator/src/edmx-parser/common/edmx-types.ts:33]: In the edmx parser I always set the value to the default `true`

File: knowledge-base/adr/0023-nullable-function-action-import-parameter.md (hunk 19 above)
Comment (FrankEssenberger) [reply knowledge-base/adr/0023-nullable-function-action-import-parameter.md:41]: removed

File: packages/generator/src/action-function-import/action.ts (hunk 3 above)
Comment (FrankEssenberger) [reply packages/generator/src/action-function-import/action.ts:14]: no it will add the `|null` a few lines later.

File: packages/generator/src/edmx-parser/v2/edmx-parser.ts (hunk 8 above)
Comment (FrankEssenberger) [reply packages/generator/src/edmx-parser/v2/edmx-parser.ts:47]: We have a isNullableParameter which does not default to true.In the generator-utls there are also other method but non seem to fist the: 
- give value
- if property not present return true.

**Review by jjtang1985 — APPROVED** [review]
LGTM

File: packages/generator/src/action-function-import/parameters-interface.spec.ts [hunk 25]
```diff
@@ -1,10 +1,13 @@
...
-import { orderBreakfast } from '../../test/test-util/data-model';
+import {
+  orderBreakfast,
+  orderBreakfastNullable
+} from '../../test/test-util/data-model';
 import { parametersInterface } from './parameters-interface';
 
 describe('parameters-interface', () => {
-  it('functionImportParametersInterface', () => {
+  it('functionImportParametersInterface not nullable', () => {
```
Comment (marikaner) [inline packages/generator/src/action-function-import/parameters-interface.spec.ts:10]: Consider giving a bit more context in the description of this and the test below.
I doubt that what is meant, is "the function `functionImportParametersInterface` is not nullable", but that is what I read.

File: packages/generator/src/action-function-import/return-type.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
+import { caps } from '@sap-cloud-sdk/util';
+import { VdmActionImport, VdmFunctionImport } from '../vdm-types';
+import { isEntityNotDeserializable } from '../edmx-to-vdm/common';
+
+function actionFunctionImportReturnType(
+  actionOrFuncitonImport: VdmActionImport | VdmFunctionImport,
```
Comment (marikaner) [inline packages/generator/src/action-function-import/return-type.ts:6]:
```suggestion
  actionOrFunctionImport: VdmActionImport | VdmFunctionImport,
```

File: packages/generator/src/edmx-parser/v2/edmx-parser.ts [hunk 27]
```diff
@@ -35,8 +35,12 @@ export function parseAssociationSets(root: any): EdmxAssociationSet[] {
...
 export function parseFunctionImports(root: any): EdmxFunctionImport[] {
-  return getPropertyFromEntityContainer(root, 'FunctionImport').map(f => ({
+  const functionImports: EdmxFunctionImport[] = getPropertyFromEntityContainer(
+    root,
+    'FunctionImport'
+  ).map(f => ({
     ...f,
     Parameter: forceArray(f.Parameter)
   }));
+  return functionImports;
```
Comment (marikaner) [inline packages/generator/src/edmx-parser/v2/edmx-parser.ts:45]: Seems like a change for debugging purposes? Consider to revert?

File: packages/generator/src/edmx-parser/v4/edmx-parser.ts [hunk 28]
```diff
@@ -69,14 +69,19 @@ export function parseActionImport(root: any): EdmxActionImport[] {
...
+): EdmxFunction[] | EdmxAction[] {
+  const actionsOrFuntions: EdmxFunction[] = getMergedPropertyWithNamespace(
+    root,
+    actionFunctionKey
+  ).map(actionOrFunction => ({
+    ...actionOrFunction,
+    Parameter: forceArray(actionOrFunction.Parameter),
+    IsBound: false
+  }));
+  return actionsOrFuntions;
```
Comment (marikaner) [inline packages/generator/src/edmx-parser/v4/edmx-parser.ts:84]: Same here, consider reverting these lines?

File: packages/generator/src/generator-utils.ts [hunk 29]
```diff
@@ -15,30 +15,65 @@ function hasCapability(object: any, capability: string) {
   return !(capability in object) || object[capability] === 'true';
 }
 
+/**
```
Comment (marikaner) [inline packages/generator/src/generator-utils.ts:18]: 👍

File: packages/generator/src/generator-utils.ts [hunk 30]
```diff
@@ -15,30 +15,65 @@ function hasCapability(object: any, capability: string) {
...
+ * Checks if the 'sap:deletable' property is present on the given input.
+ * @param entity - Object to be checked
```
Comment (marikaner) [inline packages/generator/src/generator-utils.ts:20]: We have fullstops at the end of parameter descriptions almost everywhere, for consistency it would be nice to have it the same everywhere (although this isn't part of the official docs, right).
I think this is even in the style guide.

File: packages/core/test/test-util/test-services/v4/test-service/action-imports.ts [hunk 31]
```diff
@@ -230,7 +230,7 @@ export function testActionImportSharedEntityReturnTypeCollection(
 ): Omit<
   ActionImportRequestBuilder<
     TestActionImportSharedEntityReturnTypeCollectionParameters,
-    never
+    never[]
```
Comment (marikaner) [inline packages/core/test/test-util/test-services/v4/test-service/action-imports.ts:233]: Hi, me again, about this `never[]` ;)
[`never`](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-never-type) represents a type that you cannot assign anything to, e. g. if you throw an error. What would `never[]` mean?

I think this is only used when we have no way of knowing what type to deserialize with `execute`. When deserializing by using `execute`, we throw an error, therefore the return value is `never` for `execute`.
When using `executeRaw` this type is not used.

**Review by marikaner — CHANGES_REQUESTED** [review]
LGTM, but I still think that `never[]` does not make sense. Other than that I only left minor comments.

**Review by marikaner — APPROVED** [review]

File: packages/generator/src/action-function-import/parameters-interface.spec.ts (hunk 25 above)
Comment (FrankEssenberger) [reply packages/generator/src/action-function-import/parameters-interface.spec.ts:10]: done
